### PR TITLE
PGVector, remove by ID, same impl as #1020 but on top of last dev.

### PR DESCRIPTION
**Pull request by humcqc**
same impl as #1020 but on top of last dev.

### CODE REVIEW

**Comment by humcqc** [comment]
@langchain4j , @nottyjay, this is the same impl as #1020

File: langchain4j-pgvector/src/main/java/dev/langchain4j/store/embedding/pgvector/PgVectorEmbeddingStore.java
Comment (humcqc) [inline langchain4j-pgvector/src/main/java/dev/langchain4j/store/embedding/pgvector/PgVectorEmbeddingStore.java]: Could be enhance with batch processing or in a single request if we just use a boolean in return.

Comment (langchain4j) [inline langchain4j-pgvector/src/main/java/dev/langchain4j/store/embedding/pgvector/PgVectorEmbeddingStore.java]: Shouldn't this removal be batched (in the same transaction)?

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java]:
```suggestion
    default void remove(String id) {
```
What about throwing an exception instead (in case of failures)?
Reasoning: users might not check the returned value explicitly and the problem will go unnoticed.
Exception is also more informative in terms of reason of the failure (e.g. network issue, not existing id, etc).

WDYT?

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java]:
```suggestion
    default void removeAll(List<String> ids) {
```
same here

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:99]: What about adding `removeAll()` that will remove all embeddings?
I guess this might be useful in many cases:
- original datasource was updated and one wants to simply re-ingest everything instead of figuring out what was changed and what should be removed and re-ingested again
- one might want to change the embedding model and needs to re-ingest everything

Another useful method would be `removeAll(Filter)` which is supported by some embedding stores (e.g. [Milvus](https://milvus.io/docs/insert-update-delete.md#Delete-entities), [Pinecone](https://docs.pinecone.io/guides/data/delete-data#delete-specific-records-by-metadata)), but this is probably for another PR.

WDYT?

**Review by ghost — COMMENTED** [review]
@humcqc thanks a lot!

**Comment by langchain4j** [comment]
@geoand @jmartisk could you please take a look as well? 🙏

File: langchain4j-pgvector/src/main/java/dev/langchain4j/store/embedding/pgvector/PgVectorEmbeddingStore.java
Comment (humcqc) [reply langchain4j-pgvector/src/main/java/dev/langchain4j/store/embedding/pgvector/PgVectorEmbeddingStore.java]: yes it' was what i was thinking, but wanted to keep implementation from the original committer to discuss it. 
will do all reviews today to finalize quickly.

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java
Comment (humcqc) [reply langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:99]: In my case i will have a tenant column that will be mandatory, but i will add this method for your use case.
I 100% agree with removeAll(Filter) , can do it in this pr or in another, as you want ?

Comment (humcqc) [reply langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:99]: should removeAll(Filter) replace removeAll(List<String> ids) and remove(String id) ? or you prefer keep both ?

I'm not sure id's will be very useful for delete or update  , as we will have a global metadata id to group all embeddings for a given data.

Comment (humcqc) [reply langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:99]: if we keep removeAll(Filter) and removeAll(List ids) we will have inconsistency when we call removeAll(null).
WDYT?
I propose tu use:
- remove(String id) -> remove one record, if null no record deleted
- remove(List<String>) -> remove list of records by id, if null no record deleted
- removeByFilter(Filter) -> remove filtered record, if null all record will be deleted

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:81]:
```suggestion
    default void removeAll(Collection<String> ids) {
```
WDYT?

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:94]:
```suggestion
    default void remove(Filter filter) {
```
WDYT?

**Comment by langchain4j** [comment]
@geoand @jmartisk could you please check? Does this API makes sense? I would love to include this in the release today.

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java
Comment (humcqc) [reply langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:94]: I propose to use:

remove(String id) -> remove one record, if null no record deleted
remove(Collection) -> remove list of records by id, if null no record deleted
removeByFilter(Filter) -> remove filtered records, if null all record will be deleted

or
remove(String... ids) -> remove records by id, if null no record deleted
removeByFilter(Filter) -> remove filtered records, if null all record will be deleted or removeByMetadata(Filter)

removeByMetadata(Filter) seems more adapted.

@langchain4j WDYT ?

**Comment by geoand** [comment]
LGTM, but I'll let @jmartisk comment further since he has interacted with the `EmbeddingStore` much more than I have

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java
Comment (jmartisk) [reply langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:94]: +1, `all` suggests really all

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java
Comment (jmartisk) [inline langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:92]: "Default" I assume means `filter==null`, so the docs should rather say that instead of "default"

File: langchain4j-pgvector/src/main/java/dev/langchain4j/store/embedding/pgvector/PgVectorEmbeddingStore.java
Comment (jmartisk) [inline langchain4j-pgvector/src/main/java/dev/langchain4j/store/embedding/pgvector/PgVectorEmbeddingStore.java:270]: PreparedStatement needs closing, so this needs to be wrapped in a try-with-resources too

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java
Comment (jmartisk) [reply langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:94]: On the other hand, `java.util.Collection.removeAll` also removes just some elements, so maybe we could leave it like this for consistency

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:94]: Good point, one can read this as "remove all that satisfy this filter"

Comment (humcqc) [reply langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:94]: I'm not sure about removeAll(Collection). and removeAll(Filter) 
If we keep this way, removeAll(null) will have 2 opposite behaviors:
-> https://github.com/langchain4j/langchain4j/pull/1113/files#diff-40e560d628f612b7350552586991a852f85d9a2c35f15604a4d345bc13209912R116
-> https://github.com/langchain4j/langchain4j/pull/1113/files#diff-40e560d628f612b7350552586991a852f85d9a2c35f15604a4d345bc13209912R133

@langchain4j , @jmartisk WDYT about:
removeAll() -> remove all
remove(String... ids) -> remove records by id ( one ore +), if null no record deleted
removeByMetadata(Filter) -> remove filtered records, if null all record will be deleted;

**Review by ghost — APPROVED** [review]
@humcqc thanks a lot!
I will merge it now and do some changes in a separate commit, as I can't push changes to your fork.

**Comment by humcqc** [comment]
humm, I had the review fixes in my local repo waiting for the answer on my last comment:: 
I'm not sure about removeAll(Collection). and removeAll(Filter)
If we keep this way, removeAll(null) will have 2 opposite behaviors:
-> https://github.com/langchain4j/langchain4j/pull/1113/files#diff-40e560d628f612b7350552586991a852f85d9a2c35f15604a4d345bc13209912R116
-> https://github.com/langchain4j/langchain4j/pull/1113/files#diff-40e560d628f612b7350552586991a852f85d9a2c35f15604a4d345bc13209912R133

@langchain4j , @jmartisk WDYT about:
removeAll() -> remove all
remove(String... ids) -> remove records by id ( one ore +), if null no record deleted
removeByMetadata(Filter) -> remove filtered records, if null all record will be deleted;

**Comment by langchain4j** [comment]
> humm, I had the review fixes in my local repo waiting for the answer on my last comment::

@humcqc my bad, pardon!
I am a bit in a rush, wanted it to make it in this release that I am preparing now.
BTW, I did not see a comment from you. Maybe it is pending? If yes, I can't see it...

I have [changed](https://github.com/langchain4j/langchain4j/commit/358f63c0faa4e56385f53ec89fe806496cf686f9) it so that null/empty ids/filter are not valid inputs.

**Comment by humcqc** [comment]
No pb! if the current version is ok for you it's ok for me.

> BTW, I did not see a comment from you. Maybe it is pending? If yes, I can't see it...

:) My bad so, didn't notice the pending stuff, I was a little bit frustrated to not have answers :)
How can I change this pending stuff ?

**Comment by langchain4j** [comment]
@humcqc yeah, I made this mistake many times 😆 You need to go to files changed -> review changes -> submit

**Review by humcqc — COMMENTED** [review]
:) a little bit late :)

**Pull request merged**
